### PR TITLE
Authorization Guard

### PR DESCRIPTION
**Pull request by devzgabriel**
Was also done:

- Improve error logs
- Request Logging
- Entities removed
- List filter improved

### CODE REVIEW

File: src/casl/factories/casl-ability.factory.ts [hunk 1]
```diff
@@ -1,36 +1,30 @@
...
+      can('create', ['people', 'contacts', 'addresses']);
+
+      can('update', ['people', 'contacts', 'addresses', 'users']);
+
+      can('delete', ['people', 'contacts', 'addresses']);
     }
 
-    can('update', ContactEntity, { personId: user.person.id });
-    cannot('delete', ContactEntity, { type: 'EMAIL' }); // Just to show how it works
+    // TODO: create user by self
```
Comment (devzgabriel) [inline src/casl/factories/casl-ability.factory.ts:26]: Discuss the logic of new employees:

Suggestion: leave the route open for registrations and when creating the user it must not have roles, that is, he cannot do anything in the system, until an admin gives him a position.

File: src/common/interceptors/request-log.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  ExecutionContext,
+  CallHandler,
+  Logger,
+} from '@nestjs/common';
+import { Observable } from 'rxjs';
+import { tap } from 'rxjs/operators';
+
+@Injectable()
+export class RequestLogInterceptor implements NestInterceptor {
+  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
```
Comment (SampaioLeal) [inline src/common/interceptors/request-log.interceptor.ts:14]: can we use `unknown` for this return type?
maybe typescript can infer a better one....

File: src/common/interceptors/request-log.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    const now = Date.now();
+    return next
+      .handle()
+      .pipe(
+        tap(() =>
+          logger.log(
+            `The user ${
+              user?.id || 'unknown'
+            } is making a ${method} request to route: "${url}" in ${
```
Comment (SampaioLeal) [inline src/common/interceptors/request-log.interceptor.ts:28]: > Sampaio is making this review in 14 seconds

i guess we can improve this information and make it more understandable

File: src/people/controllers/addresses.controller.ts [hunk 4]
```diff
@@ -16,8 +16,11 @@ export class AddressesController {
 
   @Auth('create', SUBJECT)
   @Post()
-  async create(@Body() createAddressDto: CreateAddressDto) {
-    return this.addressesService.create(createAddressDto);
+  async create(
+    @Param('id') id: string,
```
Comment (SampaioLeal) [inline src/people/controllers/addresses.controller.ts:20]:
```suggestion
    @Param('id') personId: string,
```

File: src/people/controllers/addresses.controller.ts [hunk 5]
```diff
@@ -16,8 +16,11 @@ export class AddressesController {
...
+    @Body() createAddressDto: CreateAddressDto,
+  ) {
+    return this.addressesService.create(id, createAddressDto);
```
Comment (SampaioLeal) [inline src/people/controllers/addresses.controller.ts:23]:
```suggestion
    return this.addressesService.create(personId, createAddressDto);
```

File: src/common/guards/roles.guard.ts [hunk 6]
```diff
@@ -1,41 +1,35 @@
...
+      AUTH_METADATA_KEY,
+      [context.getHandler()],
     );
-    return false;
+
+    const ability = this.caslAbilityFactory.createForUser(user);
+
+    const can = ability.can(action, subject);
+
+    if (!can) return false;
```
Comment (SampaioLeal) [inline src/common/guards/roles.guard.ts:31]: i guess this is unnecessary when returning `can` that is boolean

File: src/people/services/people.service.ts [hunk 7]
```diff
@@ -10,8 +10,17 @@ export class PeopleService {
   constructor(private readonly prisma: PrismaService) {}
 
   async create(createPersonDto: CreatePersonDto) {
+    const { contacts, addresses, ...newPerson } = createPersonDto;
     return this.prisma.person.create({
-      data: createPersonDto,
+      data: {
+        ...newPerson,
+        contacts: {
+          create: contacts.map((c) => c),
```
[the remark this answers is not in the document]
Comment (devzgabriel) [reply src/people/services/people.service.ts:18]: That's a type error:
Screenshot from 2022-01-13 10-22-28

**Review by SampaioLeal — APPROVED** [review]

**Pull request merged**